### PR TITLE
[master] Deprecated in favor of Salt Extension Decorator

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Adding a new decorator to indicate when a module is deprecated in favor of a Salt extension

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by Ch3LL — DISMISSED** [review]
We need to document this so users know how to properly deprecate modules in favor of Salt extensions. I'm approving this as I'm okay if we do this in a separate PR as it might require more work since we don't have any documentation related to writing  a Salt extension currently that I'm aware of.

**Review by dmurphy18 — DISMISSED** [review]
What about modules already with deprecation since moved to extensions in 3006, for example the ESXi module and it's cohorts, should they stay unchanged or be updated once this is merged ?

**Review by s0undt3ch — DISMISSED** [review]

File: doc/topics/development/salt_extensions.rst [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+.. _deprecate-modules:
+
+How do I deprecate a Salt module to a Salt extension
+----------------------------------------------------
+
+To indicate that a Salt module is being deprecated in favor of a Salt extension,
+for each Salt module include ``__deprecated__`` tuple in the module.  The tuple
+should include the version of Salt that the module will be removed, the name of the
+collection of modules that are being deprecated, and the URL where the source for
+the new extension can be found.
```
Comment (s0undt3ch) [inline doc/topics/development/salt_extensions.rst:66]:
```suggestion
should include the version of Salt that the module will be removed(should be 2 major versions from the current release. For example, the current Salt release is 3100, the deprecation version should be set to 3102), the name of the
collection of modules that are being deprecated, and the URL where the source for
the new extension can be found.
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by Ch3LL — DISMISSED** [review]

File: doc/topics/development/salt_extensions.rst [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+## Community Modules
+
+Remaining modules will be deprecated from the Salt Core codebase and community members
+will be able to continue independent maintainership if they are interested. Some plugins are
+almost exclusively maintained by external corporations – if these corporations wish for formal
+documentation outlining transfer of ownership it can be handled on a case-by-case basis. The
+community modules can be hosted either in individual or corporate source control systems,
+alternatively they can also be hosted in the community run Salt Extensions Github organization,
+that will operate like the the Salt Formulas Github organization.
+The criteria to determine which category to place modules in will follow these rules.
```
Comment (twangboy) [inline doc/topics/development/salt_extensions.rst:31]: Add a colon
```
...will follow these rules:
```

File: doc/topics/development/salt_extensions.rst [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+How do I deprecate a Salt module to a Salt extension
+----------------------------------------------------
+
+To indicate that a Salt module is being deprecated in favor of a Salt extension,
+for each Salt module include ``__deprecated__`` tuple in the module.  The tuple
+should include the version of Salt that the module will be removed, the name of the
+collection of modules that are being deprecated, and the URL where the source for
+the new extension can be found. The version should be 2 major versions from the
+next major release. For example, the next major release of Salt is 3100, the
```
Comment (twangboy) [inline doc/topics/development/salt_extensions.rst:74]: For example, `if` the next major release...

**Review by twangboy — CHANGES_REQUESTED** [review]

**Review by dmurphy18 — DISMISSED** [review]

**Review by s0undt3ch — DISMISSED** [review]

**Review by dmurphy18 — DISMISSED** [review]

File: tests/pytests/functional/utils/test_extension_deprecation.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+def salt_func_two():
+    return True
+
+
+def test_extension_deprecation():
+    """
+    this tests for a condition where an included jinja template
+    is removed from the salt filesystem, but is still loaded from
+    the cache.
+    """
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply tests/pytests/functional/utils/test_extension_deprecation.py:21]: yeah, we need to update this docstring

**Review by s0undt3ch — DISMISSED** [review]

**Review by I3urny — CHANGES_REQUESTED** [review]
Just needs a minor fix in the documentation. Looks good otherwise.

**Review by I3urny — APPROVED** [review]

**Review by dmurphy18 — APPROVED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**